storage/innobase/row/row0ins.cc

```diff
@@ -2749,6 +2749,7 @@ row_ins_clust_index_entry_low(
 #else /* BTR_CUR_HASH_ADAPT */
 			index->table->bulk_trx_id = trx->id;
 #endif /* BTR_CUR_HASH_ADAPT */
+			export_vars.innodb_bulk_operations++;
```

Comment:
Could we increment this on transaction commit? Shouldn’t we increment it also in native table-rebuilding `ALTER TABLE`?

In the current form of the patch, this status variable is not protected by anything, so there could be race conditions between concurrent threads updating it. It could be thinkable that an external tool would load data into multiple tables concurrently, and hence race conditions around this counter could lead to incorrect results (some updates of the counter being lost). If we updated this under transaction commit, we could use an existing mutex for protecting it, such as `lock_sys.wait_mutex`.

Another alternative would be to use `Atomic_relaxed` for the counter.

Reply:
@vuvova just said that incrementing the counter at the start of the transaction is acceptable.